core/src/upgrade/pending.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 Protocol Labs.
+// Copyright 2017-2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::upgrade::{InboundUpgrade, OutboundUpgrade, ProtocolName, UpgradeInfo};
+use futures::future;
+use std::iter;
+use void::Void;
+
+/// Implementation of [`UpgradeInfo`], [`InboundUpgrade`] and [`OutboundUpgrade`] that always
+/// returns a pending upgrade.
+#[derive(Debug, Copy, Clone)]
+pub struct PendingUpgrade<P> {
```

Comment:
Why do we need this to be generic?

Reply:
Are you suggesting to use a `String` instead? What if a user wants to use it with a type other than `String` that implements `ProtocolName`?

Reply:
With "user" I mean am thinking of potential users beyond the test introduced in this pull request.

Reply:
> Are you suggesting to use a `String` instead? What if a user wants to use it with a type other than `String` that implements `ProtocolName`?

`&'static str` would probably be my preference.

Reply:
`libp2p-core` has the `ProtocolName` abstraction. Thus I would suggest all of `libp2p-core` should use that abstraction. I am happy to design an alternative approach. Though I think that should happen in a different pull request.